### PR TITLE
Fix sparsemax computing the wrong threshold

### LINKED ISSUE #23425: sparsemax returns values that do not sum to 1 on all backends
`keras.ops.sparsemax` and `keras.activations.sparsemax` return incorrect values whenever
more than one element falls in the support. Sparsemax is defined as the Euclidean
projection onto the probability simplex, so its output must be non-negative and sum to 1
along the reduction axis. It currently does not.

Reproducible on the numpy, jax, tensorflow and torch backends (and by inspection in the
openvino backend, which contains the same line).

### Reproduction

```python
import numpy as np
from keras import ops

x = np.array([[0.0, 0.5, 1.0]], dtype="float32")
print(ops.sparsemax(x))
```

Actual, on every backend:

```
[[0.   0.   0.25]]     # sums to 0.25
```

Expected:

```
[[0.   0.25 0.75]]     # sums to 1.0
```

### Cause

Sparsemax (Martins & Astudillo 2016, Algorithm 1) computes the threshold from the **k-th**
cumulative sum of the descending-sorted logits, where `k` is the size of the support:

```
tau(z) = (cumsum[k] - 1) / k
```

All five backends instead sum *every* cumulative sum that falls inside the support:

```python
logits_cumsum_safe = np.where(support, logits_cumsum, 0.0)   # masks the cumulative sums
tau = (np.sum(logits_cumsum_safe, axis=axis, keepdims=True) - 1) / k
```

For the example above the sorted logits are `[1, 0.5, 0]`, the cumulative sums are
`[1, 1.5, 1.5]`, and the support is the first two entries. The correct threshold is
`(1.5 - 1) / 2 = 0.25`; the code computes `(1 + 1.5 - 1) / 2 = 0.75`.

The same line is present in:

- `keras/src/backend/numpy/nn.py`
- `keras/src/backend/jax/nn.py`
- `keras/src/backend/tensorflow/nn.py`
- `keras/src/backend/torch/nn.py`
- `keras/src/backend/openvino/nn.py`

### Scope

The result is correct only when the support holds exactly one element, because the masked
sum then collapses to `cumsum[0]`. Sampling 2000 random inputs (2–8 logits, normal with
scale 0.3 / 1.0 / 3.0), the threshold is wrong in **1381 of them (69%)**.

The existing tests all happen to fall in the one-element case: `keras/src/ops/nn_test.py`
uses `[-0.5, 0, 1, 2, 3]`, and every case in `keras/src/activations/activations_test.py`
uses `np.linspace(1, 12, ...)`. All produce one-hot output and pass either way.

### Impact

Affects both `keras.ops.sparsemax` and `keras.activations.sparsemax`, so any model using
sparsemax as an activation has been producing incorrect values.

I have a fix ready and will open a PR shortly.

### PR DESCRIPTION
**Pull request by SEPURI-SAI-KRISHNA**
Fixes #23425

## Description

`keras.ops.sparsemax` and `keras.activations.sparsemax` return incorrect values on every
backend whenever more than one element falls in the support. The output is not a
projection onto the probability simplex — it does not sum to 1.

```python
import numpy as np
from keras import ops

x = np.array([[0.0, 0.5, 1.0]], dtype="float32")
ops.sparsemax(x)
# current:  [[0.  , 0.  , 0.25]]   sum = 0.25   <- wrong, on all backends
# expected: [[0.  , 0.25, 0.75]]   sum = 1.0
```

### Cause

Sparsemax (Martins & Astudillo, 2016, Algorithm 1) defines the threshold as

```
tau(z) = (cumsum[k] - 1) / k
```

where `cumsum[k]` is the **k-th** cumulative sum of the descending-sorted logits and `k`
is the size of the support. Every backend instead sums *all* the cumulative sums that
fall in the support:

```python
logits_cumsum_safe = np.where(support, logits_cumsum, 0.0)   # masks the CUMULATIVE SUMS
tau = (np.sum(logits_cumsum_safe, axis=axis, keepdims=True) - 1) / k
```

For the example above the sorted logits are `[1, 0.5, 0]` with cumulative sums
`[1, 1.5, 1.5]` and a support of the first two entries, so `tau` should be
`(1.5 - 1) / 2 = 0.25`. The current code computes `(1 + 1.5 - 1) / 2 = 0.75`.

The identical line appears in all five backends: `numpy/nn.py`, `jax/nn.py`,
`tensorflow/nn.py`, `torch/nn.py` and `openvino/nn.py`.

### Why it was not caught

The result is only correct when the support holds exactly one element, because the masked
sum then collapses to `cumsum[0]`. Every existing test uses widely separated logits — the
`keras.ops` test uses `[-0.5, 0, 1, 2, 3]` and all the `activations` tests use
`np.linspace(1, 12, ...)` — so all of them produce one-hot output and pass either way.

Over 2000 random inputs (2–8 logits, scales 0.3/1.0/3.0), the current implementation
computes the wrong threshold in **1381 of them (69%)**.

### Fix

The support is a prefix of the descending-sorted logits, so the k-th cumulative sum is
exactly the sum of the sorted logits that fall in the support. Masking `logits_sorted`
instead of `logits_cumsum` gives the correct `tau` without needing a gather:

```python
logits_masked = np.where(support, logits_sorted, 0.0)
tau = (np.sum(logits_masked, axis=axis, keepdims=True) - 1) / k
```

One line per backend, no change in complexity or op count.

### Second fix: dtype promotion on the numpy backend (added in review)

The numpy backend also returned `float64` for every input dtype, while the other backends
preserve the input dtype:

```python
ops.sparsemax(ops.cast(x, "float16"))   # numpy backend -> float64, elsewhere -> float16
```

Three separate promotions were at work: `k` is an `int64` count and NumPy promotes
`float32 / int64` to `float64`; and the Python float literals in `np.where(..., 0.0)` and
`np.maximum(..., 0.0)` promote `bfloat16`. The constants and counts are now materialized
in the dtype of `logits`, matching the idiom already used by `relu`, `relu6` and `elu` in
the same file. `r` is cast as well, so the `support` comparison is evaluated in the input
dtype like it is on the TensorFlow and PyTorch backends, rather than in `float64`.

## Tests

- `NNOpsCorrectnessTest::test_sparsemax_dtype` — new; asserts the output dtype follows the
  input for `float16`, `float32` and `bfloat16`. Fails on the numpy backend before the
  dtype fix.
- `NNOpsCorrectnessTest::test_sparsemax` — adds `[0.0, 0.5, 1.0] -> [0.0, 0.25, 0.75]`, a
  support of size 2 whose expected output is exactly representable in `float32`. Also adds
  a property check over random inputs at three scales: the output is non-negative and sums
  to 1 along `axis`.
- `ActivationsTest::test_sparsemax` — adds the same support-of-size-2 case, so the
  `keras.activations` entry point is covered too.

Both new checks fail on numpy, jax, tensorflow and torch before the fix and pass after.

Verified against the reference algorithm on 300 random batches per backend after the fix:
max absolute deviation `1.6e-07` (float32 rounding), and every row sums to 1.

Full test files green on all four locally available backends:

| Backend | `nn_test.py` + `activations_test.py` |
|---|---|
| numpy | 449 passed, 27 skipped |
| jax | 459 passed, 17 skipped |
| tensorflow | 462 passed, 14 skipped |
| torch | 448 passed, 28 skipped |

## Notes for reviewers

- **The OpenVINO change is untested locally** — I do not have an OpenVINO runtime
  available. The edit is the direct analogue of the other four
  (`ov_opset.select(support, logits_cumsum, ...)` becomes
  `ov_opset.select(support, logits_sorted, ...)`), and `test_sparsemax` is not in
  `excluded_concrete_tests.txt`, so CI will exercise it.
- **This changes numerical output** for any model using `sparsemax`. That is the point of
  the fix, but it is worth calling out explicitly: previously saved models will now produce
  different — correct — activations.
- The numpy dtype fix was added during review, after it was flagged on a line this PR
  already touches. Casting `k` alone is not sufficient — `bfloat16` still promotes to
  `float64` through the Python float literals in `np.where` and `np.maximum` — so the
  constants are materialized in the input dtype as well.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

## AI assistance disclosure

Per the AI-Assisted Contribution Policy: I used an AI coding assistant to help find this
bug and to draft the patch, the tests and the verification reported above. I have reviewed
the change, I understand why masking the sorted logits yields the k-th cumulative sum, and
I take responsibility for the code submitted here. Review responses will be written by me.

### CODE REVIEW

File: keras/src/backend/tensorflow/nn.py [hunk 1]
```diff
@@ -197,9 +197,12 @@ def sparsemax(x, axis=-1):
     r = tf.reshape(r, r_shape)  # Reshape for broadcasting
     support = logits_sorted - (logits_cumsum - 1) / r > 0
     # Find the threshold
-    logits_cumsum_safe = tf.where(support, logits_cumsum, 0.0)
+    # `tau` is derived from the k-th cumulative sum, which is the sum of the
+    # `k` largest logits. `support` is a prefix mask, so masking the sorted
+    # logits gives that sum directly.
+    logits_masked = tf.where(support, logits_sorted, 0.0)
```
[the remark this answers is not in the document]
Comment (SEPURI-SAI-KRISHNA) [reply keras/src/backend/tensorflow/nn.py:203]: I checked this and `tf.where` does accept a Python float here — it converts the scalar to
the dtype of the tensor branch rather than to `float32`:

| tensor dtype | `tf.where(c, t, 0.0)` |
| --- | --- |
| float16 | ok, returns float16 |
| bfloat16 | ok, returns bfloat16 |
| float32 | ok, returns float32 |

No `TypeError` and no promotion. Note also that this line is not new in this PR — the
previous code was `tf.where(support, logits_cumsum, 0.0)` with the same `0.0` literal, and
sparsemax already ran on float16 and bfloat16 inputs before this change.

Leaving as is to keep the diff to the one-line fix, but happy to switch to `zeros_like` if
you would prefer it stylistically.

File: keras/src/backend/torch/nn.py [hunk 2]
```diff
@@ -210,10 +210,13 @@ def sparsemax(x, axis=-1):
...
     # Find the threshold
     k = torch.sum(support, dim=axis, keepdim=True)
-    logits_cumsum_safe = torch.where(
-        support, logits_cumsum, torch.tensor(0.0, device=logits.device)
+    # `tau` is derived from the k-th cumulative sum, which is the sum of the
+    # `k` largest logits. `support` is a prefix mask, so masking the sorted
+    # logits gives that sum directly.
+    logits_masked = torch.where(
+        support, logits_sorted, torch.tensor(0.0, device=logits.device)
     )
```
[the remark this answers is not in the document]
Comment (SEPURI-SAI-KRISHNA) [reply keras/src/backend/torch/nn.py:218]: PyTorch treats 0-dim tensors as scalars for type promotion, so this does not upcast:

| tensor dtype | `torch.where(c, t, torch.tensor(0.0))` |
| --- | --- |
| float16 | torch.float16 |
| bfloat16 | torch.bfloat16 |
| float32 | torch.float32 |

As with the TensorFlow comment, the `torch.tensor(0.0, device=...)` call is pre-existing —
this PR only changed which tensor is masked. Happy to switch to `zeros_like` if you prefer
it, but it is not fixing a promotion bug.

File: keras/src/backend/numpy/nn.py [hunk 3]
```diff
@@ -229,8 +229,11 @@ def sparsemax(x, axis=-1):
     support = logits_sorted - (logits_cumsum - 1) / r > 0
     # Find the threshold
     k = np.sum(support, axis=axis, keepdims=True)
-    logits_cumsum_safe = np.where(support, logits_cumsum, 0.0)
-    tau = (np.sum(logits_cumsum_safe, axis=axis, keepdims=True) - 1) / k
+    # `tau` is derived from the k-th cumulative sum, which is the sum of the
+    # `k` largest logits. `support` is a prefix mask, so masking the sorted
+    # logits gives that sum directly.
+    logits_masked = np.where(support, logits_sorted, 0.0)
+    tau = (np.sum(logits_masked, axis=axis, keepdims=True) - 1) / k
```
[the remark this answers is not in the document]
Comment (SEPURI-SAI-KRISHNA) [reply keras/src/backend/numpy/nn.py:240]: Good catch — this is a real bug, and I have fixed it in the latest commit.

Casting `k` alone is not quite sufficient though. It fixes float16 and float32, but
bfloat16 still ends up as float64, because the Python float literals in
`np.where(support, logits_sorted, 0.0)` and `np.maximum(logits - tau, 0.0)` promote it —
ml_dtypes types do not get NEP 50 weak-scalar treatment:

| input | cast `k` only | full fix |
| --- | --- | --- |
| float16 | float16 | float16 |
| float32 | float32 | float32 |
| bfloat16 | float64 | bfloat16 |

So the constants are now materialized in the input dtype as well, using the same
`np.array(0.0, x.dtype)` idiom that `relu`, `relu6` and `elu` already use in this file. I
also cast `r`, so the `support` comparison is evaluated in the input dtype as it is on the
TensorFlow and PyTorch backends, instead of in float64.

Added `NNOpsCorrectnessTest::test_sparsemax_dtype`, which asserts the output dtype follows
the input for float16, float32 and bfloat16, and fails on the numpy backend without this
change.

**Comment by hertschuh** [comment]
Addressed by https://github.com/keras-team/keras/pull/23469
